File: laundry/wash_wallet/packages/wash_wallet_domain/lib/src/entities/order_item_process.dart
import 'package:equatable/equatable.dart';

class OrderItemProcess extends Equatable {
  final int id;
  final int? orderItemId;
  final int? laundryServiceProcessId;
  final int? processId;
  final String? processName;
  final int? sequenceNumber;
  final String? status;
  final String? statusLabel;
  final int? employeeId;
  final String? employeeName;
  final double? qtyProcessed;
  final String? evidenceAttachment;
  final String? evidenceUrl;
  final bool hasEvidence;
  final bool canStart;
  final bool canComplete;
  final bool canWork;
  final String? actionDeniedReason;
  final DateTime? startedAt;
  final DateTime? completedAt;
  final String? formattedStartedAt;
  final String? formattedCompletedAt;
  final bool isCompleted;
  final bool isInProgress;
  final bool isPending;
  final int? processingDuration;
  final double? completionPercentage;
  final double? commissionAmount;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final String? formattedCreatedAt;
  final String? formattedUpdatedAt;

  const OrderItemProcess({
    required this.id,
    this.orderItemId,
    this.laundryServiceProcessId,
    this.processId,
    this.processName,
    this.sequenceNumber,
    this.status,
    this.statusLabel,
    this.employeeId,
    this.employeeName,
    this.qtyProcessed,
    this.evidenceAttachment,
    this.evidenceUrl,
    this.hasEvidence = false,
    this.canStart = false,
    this.canComplete = false,
    this.canWork = false,
    this.actionDeniedReason,
    this.startedAt,
    this.completedAt,
    this.formattedStartedAt,
    this.formattedCompletedAt,
    this.isCompleted = false,
    this.isInProgress = false,
    this.isPending = false,
    this.processingDuration,
    this.completionPercentage,
    this.commissionAmount,
    this.createdAt,
    this.updatedAt,
    this.formattedCreatedAt,
    this.formattedUpdatedAt,
  });

  factory OrderItemProcess.fromModel(dynamic model) {
    return OrderItemProcess(
      id: model.id,
      orderItemId: model.orderItemId,
      laundryServiceProcessId: model.laundryServiceProcessId,
      processId: model.processId,
      processName: model.processName,
      sequenceNumber: model.sequenceNumber,
      status: model.status,
      statusLabel: model.statusLabel,
      employeeId: model.employeeId,
      employeeName: model.employeeName,
      qtyProcessed: model.qtyProcessed,
      evidenceAttachment: model.evidenceAttachment,
      evidenceUrl: model.evidenceUrl,
      hasEvidence: model.hasEvidence,
      canStart: model.canStart,
      canComplete: model.canComplete,
      canWork: model.canWork,
      actionDeniedReason: model.actionDeniedReason,
      startedAt: model.startedAt,
      completedAt: model.completedAt,
      formattedStartedAt: model.formattedStartedAt,
      formattedCompletedAt: model.formattedCompletedAt,
      isCompleted: model.isCompleted,
      isInProgress: model.isInProgress,
      isPending: model.isPending,
      processingDuration: model.processingDuration,
      completionPercentage: model.completionPercentage,
      commissionAmount: model.commissionAmount,
      createdAt: model.createdAt,
      updatedAt: model.updatedAt,
      formattedCreatedAt: model.formattedCreatedAt,
      formattedUpdatedAt: model.formattedUpdatedAt,
    );
  }

  @override
  List<Object?> get props => [
    id,
    orderItemId,
    laundryServiceProcessId,
    processId,
    processName,
    sequenceNumber,
    status,
    statusLabel,
    employeeId,
    employeeName,
    qtyProcessed,
    evidenceAttachment,
    evidenceUrl,
    hasEvidence,
    canStart,
    canComplete,
    canWork,
    actionDeniedReason,
    startedAt,
    completedAt,
    formattedStartedAt,
    formattedCompletedAt,
    isCompleted,
    isInProgress,
    isPending,
    processingDuration,
    completionPercentage,
    commissionAmount,
    createdAt,
    updatedAt,
    formattedCreatedAt,
    formattedUpdatedAt,
  ];
}

File: laundry/wash_wallet_be/resources/js/types/order_item_process.ts
import {
    BaseFilters,
    BaseSortOptions,
    Employee,
    LaundryServiceProcess,
    OrderItem,
} from ".";

export interface OrderItemProcess {
    id: number;
    orderItemId: number;
    serviceProcessId: number;
    employeeId?: number | null;
    qtyProcessed: number;
    startedAt?: string | null;
    completedAt?: string | null;
    evidenceAttachment?: string | null;
    createdAt: string;
    updatedAt: string;
    orderItem?: OrderItem;
    laundryServiceProcess?: LaundryServiceProcess;
    employee?: Employee;
    evidenceUrl?: string | null;
    status?: "pending" | "in_progress" | "completed";
    processingDuration?: number | null;
    commissionAmount?: number | null;
    completionPercentage?: number;
}

/**
 * Order item process filter interface
 */
export interface OrderItemProcessFilters extends BaseFilters {
    orderItemId?: number;
    laundryServiceProcessId?: number;
    employeeId?: number;
    status?: "pending" | "in_progress" | "completed";
    hasEvidence?: boolean;
    startDate?: string;
    endDate?: string;
    startedBetween?: {
        start: string;
        end: string;
    };
    completedBetween?: {
        start: string;
        end: string;
    };
    withRelations?: boolean;
}

/**
 * Order item process sort options
 */
export interface OrderItemProcessSortOptions extends BaseSortOptions {
    column:
        | "qtyProcessed"
        | "startedAt"
        | "completedAt"
        | "createdAt"
        | "updatedAt"
        | "orderItemId"
        | "serviceProcessId"
        | "employeeId"
        | "sequence";
}

/**
 * Order item process form data
 */
export interface OrderItemProcessFormData {
    orderItemId: number;
    serviceProcessId: number;
    employeeId?: number | null;
    qtyProcessed?: number;
    startedAt?: string | null;
    completedAt?: string | null;
    evidenceAttachment?: File | string | null;
}

/**
 * Order item process statistics
 */
export interface OrderItemProcessStatistics {
    totalProcesses: number;
    completed: number;
    inProgress: number;
    pending: number;
    withEvidence: number;
    withoutEvidence: number;
    averageProcessingTime: number;
    totalQtyProcessed: number;
    totalCommission: number;
}

/**
 * Employee performance data
 */
export interface EmployeePerformance {
    employeeId: number;
    employeeName: string;
    totalProcessesCompleted: number;
    totalQtyProcessed: number;
    totalCommission: number;
    averageProcessingTime: number;
    completionRate: number;
    processes: OrderItemProcess[];
}

/**
 * Process timeline item
 */
export interface ProcessTimelineItem {
    id: number;
    processName: string;
    sequence: number;
    status: "pending" | "in_progress" | "completed";
    startedAt?: string | null;
    completedAt?: string | null;
    employeeName?: string | null;
    qtyProcessed: number;
    evidenceUrl?: string | null;
}
